repository/dimensions_repository.go

```diff
@@ -0,0 +1,132 @@
+package repository
+
+import (
+	"errors"
+	"fmt"
+	"strconv"
+
+	"github.com/ONSdigital/dp-dimension-importer/client"
+	logKeys "github.com/ONSdigital/dp-dimension-importer/common"
+	"github.com/ONSdigital/dp-dimension-importer/model"
+	"github.com/ONSdigital/go-ns/log"
+	bolt "github.com/johnnadratowski/golang-neo4j-bolt-driver"
+)
+
+const (
+	// Create a unique constraint on the dimension type value.
+	uniqueDimConstStmt = "CREATE CONSTRAINT ON (d:`%s`) ASSERT d.value IS UNIQUE"
+
+	// Create the dimension node and the HAS_DIMENSION relationship to the Instance it belongs to.
+	createDimensionAndInstanceRelStmt = "MATCH (i:`%s`) CREATE (d:`%s` {value: {value}}) CREATE (i)-[:HAS_DIMENSION]->(d) RETURN ID(d)"
+
+	stmtKey                   = "statment"
+	stmtParamsKey             = "params"
+	valueKey                  = "value"
+	dimensionsKey             = "dimensions"
+	dimensionsList            = "dimensions_list"
+	uniqueConstErr            = "Error executing unique constraint Statment"
+	uniqueConstSuccess        = "Successfully created unique constraint on dimension"
+	dimensionNilErr           = "Dimension is required but was nil"
+	dimensionInvalidErr       = "Dimension invalid: both dimension.dimension_id and dimension.value are required but were both empty"
+	dimensionIDRequiredErr    = "dimension.Dimension_ID is required but was empty"
+	dimensionValueRequiredErr = "dimension.Value is required but was empty"
+	nodeIDCastErr             = "Unexpected error while casting NodeID to int64"
+	insertDimValidationMsg    = "InsertDimension: dimension failed validation"
+	errExecutingStatment      = "Error executing statement"
+	errReturningRows          = "Error return query rows"
+	uniqueConstraintErr       = "Unexected error while attempting to create unique Dimension ID constaint."
+	insertDimErr              = "Unexpected error while attempting to create dimension"
+)
+
+type NeoDatabase interface {
+	Query(query string, params map[string]interface{}) (*client.NeoRows, error)
+	ExecStmt(query string, params map[string]interface{}) (bolt.Result, error)
+}
+
+type DimensionRepository struct {
+	ConstraintsCache map[string]string
+	Neo              NeoDatabase
+}
+
+func (repo DimensionRepository) Insert(i *model.Instance, d *model.Dimension) (*model.Dimension, error) {
+	if _, exists := repo.ConstraintsCache[d.DimensionID]; !exists {
+
+		if err := repo.createUniqueConstraint(d); err != nil {
+			log.ErrorC(uniqueConstraintErr, err, nil)
+			return nil, err
+		}
+		repo.ConstraintsCache[d.DimensionID] = d.DimensionID
+		i.AddDimension(d)
+	}
+
+	if d, err := repo.insertDimension(i, d); err != nil {
+		log.Debug(insertDimErr, log.Data{
+			logKeys.DimensionID:  d.DimensionID,
+			logKeys.ErrorDetails: err.Error(),
+		})
+		return nil, err
+	}
+	return d, nil
+}
+
+func (repo DimensionRepository) createUniqueConstraint(d *model.Dimension) error {
+	// validate d.
```

Comment:
Is this a commented out method call? Or actually a comment - if comment, could be more descriptive :)